beacon-chain/sync/initial-sync/blocks_fetcher.go

```diff
@@ -0,0 +1,337 @@
+func (f *blocksFetcher) iter() <-chan *fetchRequestResponse {
+	return f.receivedFetchResponses
+}
+
+// loop is a main fetcher loop, listens for incoming requests/cancellations, forwards outgoing responses
+func (f *blocksFetcher) loop() {
+	defer close(f.receivedFetchResponses)
+
+	randGenerator := rand.New(rand.NewSource(time.Now().Unix()))
+	highestFinalizedSlot := helpers.StartSlot(f.highestFinalizedEpoch() + 1)
+	maxPeersToSync := params.BeaconConfig().MaxPeersToSync
+
+	for {
+		select {
+		case <-f.ctx.Done():
+			// Upstream context is done.
+			log.Warn("Upstream context canceled. Blocks fetcher is stopped.")
+			return
+		case <-f.quit:
+			// Terminating abort all operations.
+			return
+		case req := <-f.requests:
+			root, finalizedEpoch, peers := f.p2p.Peers().BestFinalized(maxPeersToSync, helpers.SlotToEpoch(f.headFetcher.HeadSlot()))
+			log.WithFields(logrus.Fields{
+				"request":        req,
+				"finalizedEpoch": finalizedEpoch,
+				"numPeers":       len(peers),
+			}).Debug("Block fetcher receives request")
+
+			if len(peers) == 0 {
+				log.Warn("No peers available, waiting for reconnect...")
+				time.Sleep(refreshTime)
+				continue
+			}
+
+			if len(peers) >= flags.Get().MinimumSyncPeers {
+				highestFinalizedSlot = helpers.StartSlot(finalizedEpoch + 1)
+			}
+
+			// Short circuit start far exceeding the highest finalized epoch in some infinite loop.
+			if req.start > highestFinalizedSlot {
+				err := errors.Errorf("requested a start slot of %d which is greater than the next highest slot of %d", req.start, highestFinalizedSlot)
+				log.WithError(err).Debug("Block fetch request failed")
+				f.receivedFetchResponses <- &fetchRequestResponse{
+					params: req,
+					err:    err,
+				}
+				continue
+			}
+
+			// Shuffle peers to prevent a bad peer from
+			// stalling sync with invalid blocks.
+			randGenerator.Shuffle(len(peers), func(i, j int) {
+				peers[i], peers[j] = peers[j], peers[i]
+			})
+
+			// TODO(4815): Consider splitting peers into sets (when there are many of them),
+			// so that block fetching is for lesser chunks and is less affected by some slow peer
+			go func(root []byte, finalizedEpoch, start, count uint64, peers []peer.ID) {
+				resp, err := f.processFetchRequest(root, finalizedEpoch, start, 1, count, peers)
+				if err != nil {
+					log.WithError(err).Debug("Block fetch request failed")
+					f.receivedFetchResponses <- &fetchRequestResponse{
+						params: req,
+						err:    err,
+					}
+					return
+				}
+
+				f.receivedFetchResponses <- &fetchRequestResponse{
+					params: req,
+					blocks: resp,
+				}
+			}(root, finalizedEpoch, req.start, req.count, peers)
+		}
+	}
+}
+
+// scheduleRequest adds request to incoming queue.
+// Should be non-blocking, as actual request processing is done asynchronously.
+func (f *blocksFetcher) scheduleRequest(req *fetchRequestParams) {
+	go func() { // non-blocking, we can re-throw requests within consuming method
+		select {
```

Comment:
do we actually need a select here ? it seems a bit redundant . we can check if quit is a closed channel at the start and if not pass the request to our `requests` channel. Since that is non-buffered maybe enclose it in a separate go-routine.